### PR TITLE
Ripple blockchain & wallet services

### PR DESCRIPTION
**Pull request by tamadamas**

### CODE REVIEW

**Comment by ysv** [comment]
@rxx Please rebase

File: config/templates/seed/blockchains.yml.erb [hunk 1]
```diff
@@ -71,10 +71,21 @@
...
   server:            http://user:password@127.0.0.1:5005 # Public Ripple node endpoint.
-  height:            11773777                            # Initial block number from which sync will be started.
+  height:            11947454                            # Initial block number from which sync will be started.
   min_confirmations: 6                                   # Minimal confirmations needed for withdraw and deposit confirmation.
   explorer:
     address:         https://bithomp.com/explorer/#{address}
     transaction:     https://bithomp.com/explorer/#{txid}
   status:            active
+
+- key:               xrp-mainnet
```
Comment (ysv) [inline config/templates/seed/blockchains.yml.erb]: @rxx Remove mainnet please
It's just example for eth only

File: spec/factories/wallet.rb [hunk 2]
```diff
@@ -42,6 +42,31 @@
       uri                'http://127.0.0.1:18332'
       secret             'changeme'
     end
-  end
 
-end
\ No newline at end of file
+    trait :xrp_hot do
```
Comment (ysv) [inline spec/factories/wallet.rb]: Can you check if we use warm wallets for xrp and eth if we don't remove them please

File: spec/resources/ripple-data/transactions.json [hunk 3]
```diff
@@ -0,0 +1,2526 @@
+{
```
Comment (ysv) [inline spec/resources/ripple-data/transactions.json]: I understand your point that ripple doesn't have blocks
But still please use name convention for this file as we do for other currencies
`start_block-end_block.json`

**Review by ysv — CHANGES_REQUESTED** [review]

**Comment by ysv** [comment]
@rxx Also rebase please

File: spec/factories/wallet.rb (hunk 2 above)
Comment (tamadamas) [reply spec/factories/wallet.rb]: we use `eth_warm` at `models/wallet_spec.rb`

File: lib/wallet_client/rippled.rb [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+                                                       .yield_self do |result|
+        result.slice('key_type', 'master_seed', 'master_seed_hex',
+                      'master_key', 'public_key', 'public_key_hex')
+              .merge(address: normalize_address(result.fetch('account_id')), secret: secret)
+              .symbolize_keys
+      end
+    end
+
+    def normalize_address(address)
+      address
```
Comment (ysv) [inline lib/wallet_client/rippled.rb]: We need to remove destination tag when we normalize address for example when we use bitgo as hot wallet
```
{"result"=>
  {"error"=>"invalidParams",
   "error_code"=>31,
   "error_message"=>"Invalid field 'tx_json.Destination'.",
   "request"=>
    {"command"=>"sign",
     "fee_mult_max"=>1000,
     "secret"=>"/P9jbbFEPt3FRtY)*DT!bDomY&aawSCKgQ89+V@nD488y4zgIgAy5$4+GVHFz)YO",
     "tx_json"=>
      {"Account"=>"rM76eUvB3RCzcF5q4Rj2Y4Y36VFbFt4csB",
       "Amount"=>"100000000",
       "Destination"=>"rfUKVVX5KJQ2FsWYrjS6JMedz3jdjPhRPZ?dt=32",
       "DestinationTag"=>32,
       "TransactionType"=>"Payment"}},
   "status"=>"error"}}
```

**Review by ysv — CHANGES_REQUESTED** [review]

File: lib/wallet_client/rippled.rb (hunk 4 above)
Comment (ysv) [reply lib/wallet_client/rippled.rb]: @rxx Also can you check if we can perform some withdrawals to addresses with tag
``` 
       ...
       "Destination"=>"rfUKVVX5KJQ2FsWYrjS6JMedz3jdjPhRPZ",
       "DestinationTag"=>32,
       ...
```

File: app/services/blockchain_service/ripple.rb [hunk 5]
```diff
@@ -1,9 +1,103 @@
...
+
+        deposits.map { |d| d[:txid] }.join(',').tap do |txids|
+          Rails.logger.info { "Deposit transactions in ledger #{ledger_index}: #{txids}" }
+        end
+
+        withdrawals.map { |d| d[:txid] }.join(',').tap do |txids|
+          Rails.logger.info { "Withdraw transactions in ledger #{ledger_index}: #{txids}" }
+        end
+
+        update_or_create_deposits!(deposits)
```
Comment (ysv) [inline app/services/blockchain_service/ripple.rb]: https://github.com/rubykube/peatio/blob/master/app/services/blockchain_service/ethereum.rb#L30

File: lib/wallet_client/rippled.rb [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+          end
+        end
+      end
+    end
+
+    def generate_destination_tag
+      begin
+        # Reserve destination 1 for system purpose
+        tag = SecureRandom.random_number(10**9) + 2
+      end while PaymentAddress.where(currency_id: :xrp)
```
Comment (ysv) [inline lib/wallet_client/rippled.rb]: we don't want to have any models in lib

File: lib/wallet_client/rippled.rb [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+    def latest_block_number
+      Rails.cache.fetch :latest_ripple_ledger, expires_in: 5.seconds do
+        response = json_rpc(:ledger, [{ "ledger_index": 'validated' }])
+        response.dig('result', 'ledger_index').to_i
+      end
+    end
+
+    def create_address!(options = {})
+      {
+        address: "#{wallet.address}?dt=#{generate_destination_tag}",
```
Comment (ysv) [inline lib/wallet_client/rippled.rb]: you can pass reserved destination tags in options

**Review by ysv — CHANGES_REQUESTED** [review]

**Review by ysv — APPROVED** [review]

**Pull request merged**